clients/src/main/java/org/apache/kafka/clients/consumer/internals/GroupCoordinator.java

```diff
@@ -0,0 +1,635 @@
+
+                SyncGroupRequest request = new SyncGroupRequest(groupId, generation,
+                        memberId, serializeAssignment(groupAssignment));
+                log.debug("Issuing leader SyncGroup ({}: {}) to coordinator {}", ApiKeys.SYNC_GROUP, request, this.coordinator.id());
+                return sendSyncGroupRequest(request);
+            } catch (RuntimeException e) {
+                return RequestFuture.failure(e);
+            }
+        } else {
+            // send follower's sync group with an empty assignment
+            SyncGroupRequest request = new SyncGroupRequest(groupId, generation,
+                    memberId, Collections.<String, ByteBuffer>emptyMap());
+            log.debug("Issuing follower SyncGroup ({}: {}) to coordinator {}", ApiKeys.SYNC_GROUP, request, this.coordinator.id());
+            return sendSyncGroupRequest(request);
+        }
+    }
+
+    private RequestFuture<S> sendSyncGroupRequest(SyncGroupRequest request) {
+        if (coordinatorUnknown())
+            return RequestFuture.coordinatorNotAvailable();
+        return client.send(coordinator, ApiKeys.SYNC_GROUP, request)
+                .compose(new SyncGroupRequestHandler());
+    }
+
+    private class SyncGroupRequestHandler extends CoordinatorResponseHandler<SyncGroupResponse, S> {
+
+        @Override
+        public SyncGroupResponse parse(ClientResponse response) {
+            return new SyncGroupResponse(response.responseBody());
+        }
+
+        @Override
+        public void handle(SyncGroupResponse syncResponse,
+                           RequestFuture<S> future) {
+            short errorCode = syncResponse.errorCode();
+            if (errorCode == Errors.NONE.code()) {
+                try {
+                    Object assignment = groupProtocol.stateSchema().read(syncResponse.memberAssignment());
+                    future.complete(groupProtocol.stateSchema().validate(assignment));
+                    sensors.syncLatency.record(response.requestLatencyMs());
+                } catch (SchemaException e) {
+                    future.raise(e);
+                }
+            } else {
+                GroupCoordinator.this.rejoinNeeded = true;
+                future.raise(Errors.forCode(errorCode));
+            }
+        }
+    }
+
+    /**
+     * Discover the current coordinator for the group. Sends a GroupMetadata request to
+     * one of the brokers. The returned future should be polled to get the result of the request.
+     * @return A request future which indicates the completion of the metadata request
+     */
+    private RequestFuture<Void> sendGroupMetadataRequest() {
+        // initiate the group metadata request
+        // find a node to ask about the coordinator
+        Node node = this.client.leastLoadedNode();
+        if (node == null) {
+            // TODO: If there are no brokers left, perhaps we should use the bootstrap set
```

Comment:
Can probably remove this comment since #290 is addressing this as part of KAFKA-1843.